Fixes #32370 - detect CentOSStream os.distro.id

lzap opened the pull request:
Puppet does report CentOSStream in os.distro.id instead of CentOS and this is the recommended way of detection. This will allow working with both OSes for easier transition to Stream.

For more info read https://community.theforeman.org/t/provisioned-hosts-with-centos8-stream-change-os-after-first-puppet-run/23157

lzap committed:
Fixes #32370 - detect CentOSStream os.distro.id

ekohl commented on test/static_fixtures/facts/centos8-legacy-facter-2.5.json:
```diff
@@ -0,0 +1,118 @@
+{
```
Please rename this. The CentOS project prefers CentOS Linux vs CentOS Stream to differentiate the two.

ekohl reviewed (changes requested):
I am not entirely sure this is really the way to go. CentOS Linux (8) will be EOL at the end of the year. At that point, do we want to keep using the name Stream or should it just be CentOS then?

Another thing to consider is what other fact parsers do. While they're currently not in core, it could cause flip flopping between the two. For example, will subscription-manager in Katello continue to update it to CentOS and Puppet back to CentOSStream? Both are constantly checking in which will cause a storm of updates.

lzap commented:
> I am not entirely sure this is really the way to go. CentOS Linux (8) will be EOL at the end of the year. At that point, do we want to keep using the name Stream or should it just be CentOS then?

I do realize that, however users struggle and I cannot figure out a different solution. We cannot rename CentOS to "CentOS Linux" that will break everyones API calls or CLI scripts.

> Another thing to consider is what other fact parsers do. While they're currently not in core, it could cause flip flopping between the two. For example, will subscription-manager in Katello continue to update it to CentOS and Puppet back to CentOSStream? Both are constantly checking in which will cause a storm of updates.

Yes, we do need a real Common Fact Model where Foreman would normalize key facts coming from different facters. I actually also grabbed JSON dumps of Ansible before and after CentOS conversion to Stream so I can take a look into that too if needed. For the record the only difference in that case is:

```diff
[lzap@nuc centos-stream-facter.tar.gz-bKmxEUx8]$ diff centos8-legacy-ansible-2.9.json centos8-stream-ansible-2.9.json
167c167
<         "ansible_distribution_version": "8.3",
---
>         "ansible_distribution_version": "8",
```

ezr-ondrej commented:
I don't feel like the one to decide here, but this should really get in. We need to start supporting centos stream and we should even cherry-pick this to 2.5.

lzap commented:
Ondro, I do not think we have some disagreement here. This needs to be fixed and we are already late. I am open to other suggestions, but currently, it is broken and this patch offers one solution - if you see that we can do better just propose what to do and I will do it.

I renamed the files as requested, rebased.

ekohl commented:
Giving a summary of the current status.

Background: Previously we had CentOS. Now the CentOS project prefers this:
* CentOS is the project
* CentOS Linux is the RHEL rebuild
* CentOS Stream is the upstream for RHEL

So we should align to that. The challenge is that we need to update all fact parsers. I know @domitea is working on moving all fact parsers into core. Should wait for that to complete before moving forward? Otherwise we risk going incompatibilities between fact parsers.

Note that I am also proposing we rename CentOS (as we have it today) to CentOS Linux. That will disambiguate it.

lzap commented:
Let's get these merged then and then proceed with this PR:

* https://github.com/theforeman/foreman/pull/8512
* https://github.com/theforeman/foreman/pull/8511
* https://github.com/theforeman/foreman/pull/8508

ezr-ondrej commented:
We can proceed here @lzap  I believe.

ekohl commented:
Related discussion: https://community.theforeman.org/t/correctly-identify-almalinux/24879/23

lzap commented:
This is being replaced with https://github.com/theforeman/foreman/pull/8870 I do not have time for this anyway.
